core/templates/filters/translate.pipe.ts

```diff
@@ -0,0 +1,149 @@
+ * The object following the pipe, i.e. {x: 'val'}, is another argument to the
+ * transform function. This object is called params or interpolationParams.
+ * Note: For every "| translate" angular encounters in the html, it creates a
+ * seperate instance of this pipe.
+ */
+
+@Pipe({
+  name: 'translate',
+  // This option is required to update the interpolatedValue when promises are
+  // resolved.
+  pure: false
+})
+export class TranslatePipe implements PipeTransform, OnDestroy {
+  interpolatedValue: string = '';
+  /**
+   * Variable to keep track of the last key sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastKey: string;
+  /**
+   * Variable to keep track of the last param sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastParams: Object;
+  onLangChange: Subscription;
+
+  constructor(
+    private translateService: TranslateService,
+    private ref: ChangeDetectorRef,
+    private utilsService: UtilsService
+  ) {}
+
+  /**
+   * Updates the value of interpolatedValue.
+   * @param {string} key - key for i18n translation
+   * @param {Object} interpolateParams - Params for interpolation
+   */
+  updateInterpolatedValue(key: string, interpolateParams?: Object): void {
+    this.interpolatedValue = this.translateService.getInterpolatedString(
+      key, interpolateParams);
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+
+    // Storing the params to check if the params are same when the transform is
+    // invoked again.
+    this.lastParams = interpolateParams;
+    this.ref.markForCheck();
+  }
+
+  /**
+   * @param {string} key - key for i18n
+   * @param {Object} params - params for interpolation
+   * @returns {string} - Interpolated I18n value
+   */
+  transform(key: string, params?: Object | undefined): string {
+    if (!key || !key.length) {
+      return key;
+    }
+
+    // If the key and params are same, return the last stored value.
+    if (
+      key === this.lastKey &&
+      this.utilsService.isEquivalent(params, this.lastParams)) {
+      return this.interpolatedValue;
+    }
+
+    let interpolateParams: Object;
+    if (this.utilsService.isDefined(params) && !Array.isArray(params)) {
+      interpolateParams = params;
+    }
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+
+    // Storing the params to check if the params are same when the transform is
+    // invoked again.
+    this.lastParams = params;
+
+    // Update the interpolated value.
+    this.updateInterpolatedValue(key, interpolateParams);
+
+    // If there is a subscription to onLangChange, unsubscribe.
```

Comment:
Perhaps explain why you're doing this (I mean, it reads as though you are unsubscribing and then re-subscribing, so the reader might wonder why).

Also, if onLangChange is always undefined after this then why do you need the next if condition?